File: src/model/pieces/PieceImplBomb.java
package model.pieces;

import model.enums.BombPieceType;

public final class PieceImplBomb extends Piece {

    public final static char NONE_PIECE  = ' ';
    public final static char WHITE_PIECE = '●';
    public final static char BLACK_PIECE = '○';
    public final static char BARRIER_PIECE  = '#';
    public final static char CRATER_PIECE  = '@';
    public final static char VALID_MOVE  = '·';

    private BombPieceType type;

    public PieceImplBomb(){
        super();
        type = BombPieceType.NORMAL;
    }

    @Override
    public void setPiece(Piece piece) {
        if(!(piece instanceof PieceImplBomb pieceImplMonochrome)) {
            throw new IllegalArgumentException();
        }
        setPlayer(pieceImplMonochrome.getPlayer());
        setType(pieceImplMonochrome.getType());
    }

    public void setType(BombPieceType type) {
        this.type = type;
    }

    public BombPieceType getType() {
        return type;
    }

    /**
     * Get the pixel form of this piece
     *
     * @return Pixel
     */
    @Override
    public int getCode() {
        return switch (this.getType()) {
            case NORMAL -> switch (this.getPlayer()) {
                case WHITE -> WHITE_PIECE;
                case BLACK -> BLACK_PIECE;
                default -> NONE_PIECE;};
            case CRATER -> CRATER_PIECE;
            case BARRIER -> BARRIER_PIECE;
        };
    }
}

File: src/model/pieces/PieceImplBomb.kt
package model.pieces

import model.enums.BombPieceType
import model.enums.Player

class PieceImplBomb : Piece() {
    @JvmField
    var type: BombPieceType

    init {
        type = BombPieceType.NORMAL
    }

    override fun setPiece(piece: Piece?) {
        require(piece is PieceImplBomb)
        player = piece.player
        this.type = piece.type
    }

    override val code: Int
        /**
         * Get the pixel form of this piece
         *
         * @return Pixel
         */
        get() = when (type) {
                    BombPieceType.NORMAL -> when (player) {
                        Player.WHITE -> WHITE_PIECE
                        Player.BLACK -> BLACK_PIECE
                        else -> NONE_PIECE
                    }

                    BombPieceType.CRATER -> CRATER_PIECE
                    BombPieceType.BARRIER -> BARRIER_PIECE
                }.code

    companion object {
        const val NONE_PIECE: Char = ' '
        const val WHITE_PIECE: Char = '●'
        const val BLACK_PIECE: Char = '○'
        const val BARRIER_PIECE: Char = '#'
        const val CRATER_PIECE: Char = '@'
        const val VALID_MOVE: Char = '·'
    }
}